### ISSUE TITLE
Add correct size checking to IAM ManagedPolicy resource types

### ISSUE THREAD

**Issue by Enrique-MD** [aws-cloudformation/cfn-lint#2221]
*cfn-lint version: (cfn-lint 0.58.1)*

*Description of issue.*

The PolicyDocument specified in an IAM Managed Policy has a maximum length of 6,144 characters (after converted to JSON and minified).
https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_iam-limits.html

When creating through CloudFormation an AWS::IAM::ManagedPolicy and using conditionals to leverage a 3 level environment creation differentiation, is throwing an error **E3502 PolicyDocument JSON text cannot be longer than 6144 characters**.

That happens because the policy defined in CloudFormation with the 3 level conditionals exceeds the 6144 characters limit, but when it is created it only uses 1 of the paths in the conditionals so the size is not really over 6144 characters.

This is the sample code in CloudFormation:
```
  ResourcePolicy:
    Type: AWS::IAM::ManagedPolicy
    Properties:
      ManagedPolicyName: !Sub
        - ResourcePolicy
      PolicyDocument:
        Version: '2012-10-17'
        Statement:
          - Sid: LambdaExecutionLogs
            Effect: Allow
            Action:
              - logs:CreateLogGroup
              - logs:CreateLogStream
              - logs:PutLogEvents
            Resource:
              !Split
                - ","
                - !Join
                    - ","
                    - - !If
                        - IsDev
                        - !Join
                          - ","
                          - 
                            - !Sub
                              arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/lambda/lambdadev1-${EnvNameCamelCase}
                            - !Sub
                              arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/lambda/lambdadev1-:*
                            - !Sub
                              arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/lambda/lambdadev2
                            - !Sub
                              arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/lambda/lambdadev2:*
                        - !Ref AWS::NoValue
                      - !If
                        - IsUat
                        - !Join
                          - ","
                          - 
                            - !Sub
                              arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/lambda/lambdauat1
                            - !Sub
                              arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/lambda/lambdauat1:*
                            - !Sub
                              arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/lambda/lambdauat2
                            - !Sub
                              arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/lambda/lambdauat2:*
                            - !Sub
                              arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/lambda/lambdauat3
                            - !Sub
                              arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/lambda/lambdauat3:*
                        - !Ref AWS::NoValue
                      - !If
                        - IsProd
                        - !Join
                          - ","
                          - 
                            - !Sub
                              arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/lambda/lambdaprod
                            - !Sub
                              arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/lambda/lambdaprod:*
                        - !Ref AWS::NoValue
          - Sid: ........
          . . .
```

So the missing feature is to be able to check the size of the policy which is going to be finally created, not the source CloudFormation code, so in that regard it needs to evaluate the parameter and resolve the conditionals to end up with the final resource code to be created.

**Comment by PatMyron** [comment]
Unfortunately probably approaching diminishing returns handling all those nested CloudFormation functions specifically for this IAM rule in my opinion

**Recommended workaround with [resource-level ignores](https://github.com/aws-cloudformation/cfn-python-lint#resource-based-metadata):**

```yaml
Resources:
  ResourcePolicy:
    Type: AWS::IAM::ManagedPolicy
    Metadata:
      cfn-lint:
        config:
          ignore_checks:
            - E3502
```

**Comment by Enrique-MD** [comment]
Thanks for the workaround @PatMyron, I had removed the E3502 checker in the whole CloudFormation template (within my pipeline linter stage) but that workaround fits better just for the particular policy.
However, I believe it would be nice to have such feature in the long term but I know it is not straightforward.

**Comment by kddejong** [comment]
We do resolve a lot of conditions but going into splits and joins, etc is not handled at this time.  At this point we will just blank them out so we don't cause false positives.   I would assume your joins into splits is used for an example as there would be better ways to write that.

**Comment by Enrique-MD** [comment]
Hi @kddejong, the resoning behind having Splits and Joins is because I need to get as output from each conditional a list of values and the only way I found (supported in CloudFormation) is this one, in other words, the resource should be a list of values depending on the conditional evaluated.

**Comment by kddejong** [comment]
ahh I see it now. Yea that may be the best way considering a bunch of a scenarios and merging a list.  If it wasn't for the `Fn::Sub` I would suggest a mapping.  If what you are doing is more like a switch you could also go with something like this.  Its really just a matter of preference though.  So pick what makes sense for you.

```yaml
ResourcePolicy:
  Type: AWS::IAM::ManagedPolicy
  Properties:
    Resource:
      !If
        - IsDev
        - 
          - !Sub arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/lambda/lambdadev1-${EnvNameCamelCase}
          - !Sub arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/lambda/lambdadev1-:*
          - !Sub arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/lambda/lambdadev2
          - !Sub arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/lambda/lambdadev2:*
        - !If
          - IsUat
          - 
            - !Sub arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/lambda/lambdauat1
            - !Sub arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/lambda/lambdauat1:*
            - !Sub arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/lambda/lambdauat2
            - !Sub arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/lambda/lambdauat2:*
            - !Sub arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/lambda/lambdauat3
            - !Sub arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/lambda/lambdauat3:*
          - !If
            - IsProd
            - 
              - !Sub arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/lambda/lambdaprod
              - !Sub arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/lambda/lambdaprod:*
            - !Ref AWS::NoValue
```

**Comment by Enrique-MD** [comment]
Oh that's very helpful, I didn't notice chaining one If with the other would solve the list issue. Thanks for that.
